content/blog/performant-code/index.Rmd

```diff
@@ -0,0 +1,526 @@
+
+```{r bm-tibble}
+bench::mark(
+  tibble = tibble(a = 1:2, b = 3:4),
+  new_tibble_df_list = new_tibble(df_list(a = 1:2, b = 3:4), nrow = 2),
+  new_tibble_list = new_tibble(list(a = 1:2, b = 3:4), nrow = 2)
+) %>% 
+  select(expression, median)
+```
+
+Note that `new_tibble()` *will not check the lengths of its inputs.* Carry out simple recycling yourself, and be sure to use the `nrow` argument to get basic length checks.
+
+Some examples of changes like this made to tidymodels packages: [tidymodels/parsnip#945](https://github.com/tidymodels/parsnip/pull/932), [tidymodels/parsnip#934](https://github.com/tidymodels/parsnip/pull/934), [tidymodels/parsnip#929](https://github.com/tidymodels/parsnip/pull/929), [tidymodels/parsnip#923](https://github.com/tidymodels/parsnip/pull/923), [tidymodels/parsnip#902](https://github.com/tidymodels/parsnip/pull/902), [tidymodels/dials#277](https://github.com/tidymodels/dials/pull/277), and [tidymodels/tune#637](https://github.com/tidymodels/tune/pull/637).
+
+### Becoming join-critical
+
+Two truths:
+
+- dplyr joins are a remarkably safe and powerful way to synthesize data sources.
+
+- One ought to ask themselves "does this really need to be a join?" when combining data sources in package code.
+
+Some ways to intuit whether joins are necessary:
+
+- Does the complete information I need live somewhere farther up the call stack or in a package environment?
+
+    - If so, could I pass it as an argument or retrieve it from a package environment?
+
+    - If not, could it? i.e. if the elements of the join are not indices of a loop, or are indices that could easily be subsetted on after the fact, do the join just once higher up in the call stack.
+
+- Am I using the complete outputted join result or just a portion? If I end up only making use of column names, or values in one column, or pairings between two columns, I may be able to instead use `$.tbl_df` or `[.tbl_df`.
+
+
+As an example, imagine we have another tibble that tells us additional information about the `make_model`s that I drove in high school:
+
+```{r simons-car}
+simons_cars <- 
+  tibble(
+    make_model = "Honda Civic",
+    vibey = "Yes"
+  )
+
+simons_cars
+```
+
+I _could_ use a join to subset down to only the cars I drove in high school and append this additional information like so:
+
+```{r right-join}
+right_join(mtcars_tbl, simons_cars, "make_model")
+```
+
+Another way to express this, though, is to keep rows that match the `make_model` in `simons_cars` and then add that `vibey` column on after:
+
+```{r supplement-simons-cars}
+supplement_simons_cars <- function() {
+  res <-   
+    vec_slice(
+      mtcars_tbl, 
+      mtcars_tbl$make_model %in% simons_cars$make_model
+    )
+  
+  res$vibey <- simons_cars$vibey
```

Comment:
I don't think this is exactly right because it relies on the ordering between the two being the same and on `simons_cars` being 100% represented in `mtcars_tbl`

I'd be tempted to refactor this a little into something that uses `vec_match()` and `vec_cbind()`, and that shows how the `inner_join()` is being mimic-ed by dropping unmatched rows.

This doesn't quite fit your story, but maybe you can make a columns like `dream = c(FALSE, TRUE)` to be like "car i actually owned vs car i owned in a dream" or something similar.

``` r
library(tidyverse)
library(vctrs)

mtcars_tbl <- as_tibble(mtcars, rownames = "make_model")

simons_cars <- 
  tibble(
    make_model = c("Honda Civic", "Honda Accord"),
    vibey = c("Yes", "No")
  )

# Pull in the details about my cars
simons_cars %>%
  inner_join(mtcars_tbl, by = join_by(make_model))
#> # A tibble: 1 × 13
#>   make_model  vibey   mpg   cyl  disp    hp  drat    wt  qsec    vs    am  gear
#>   <chr>       <chr> <dbl> <dbl> <dbl> <dbl> <dbl> <dbl> <dbl> <dbl> <dbl> <dbl>
#> 1 Honda Civic Yes    30.4     4  75.7    52  4.93  1.62  18.5     1     1     4
#> # ℹ 1 more variable: carb <dbl>

# If you expect that each of Simon's cars has exactly 0 or 1 match in
# `mtcars_tbl`, then you can do this instead
loc <- vec_match(simons_cars$make_model, mtcars_tbl$make_model)
loc
#> [1] 19 NA

# Keep only the matches
loc_simons <- which(!is.na(loc))
loc_mtcars <- vec_slice(loc, !is.na(loc))

# Drop duplicated join column
mtcars_tbl_join <- mtcars_tbl[setdiff(names(mtcars_tbl), "make_model")]

vec_cbind(
  vec_slice(simons_cars, loc_simons),
  vec_slice(mtcars_tbl_join, loc_mtcars)
)
#> # A tibble: 1 × 13
#>   make_model  vibey   mpg   cyl  disp    hp  drat    wt  qsec    vs    am  gear
#>   <chr>       <chr> <dbl> <dbl> <dbl> <dbl> <dbl> <dbl> <dbl> <dbl> <dbl> <dbl>
#> 1 Honda Civic Yes    30.4     4  75.7    52  4.93  1.62  18.5     1     1     4
#> # ℹ 1 more variable: carb <dbl>
```

<sup>Created on 2023-04-14 with [reprex v2.0.2.9000](https://reprex.tidyverse.org)</sup>

Reply:
I still think your comment at the end applies (about how if it got more complex then you may have some issues), but I think this is a little closer to "what I'd actually do" when porting code to vctrs